Batch/Serial Number entry modal

pratu16x7 opened the pull request:
Depends on frappe/frappe#3298

Triggers on selecting item_code after warehouse.

- Adds initial batches by default, and prefills qty sorting batches by oldest.
- Serialized items have a link to text-area adder (No smart behaviour related to item qty).
- Modal item quantity updates as batches or serial numbers are added; setting it manually doesn't do anything (flow-breaking?)
![batch2](https://cloud.githubusercontent.com/assets/5196925/26346216/212c5b84-3fc3-11e7-9793-f05f27033340.gif)
Though the new row values are set, they don't seem to be triggered(indicator stays orange) even after `refresh_field("items")`. Hints?


For fetching available quantities, the modal has a prerequisite of setting the warehouse before item_code:
![batch1](https://cloud.githubusercontent.com/assets/5196925/26346198/14589846-3fc3-11e7-88dc-08da910c0833.gif)
As such, delivery note implementation turned out to be clunky; the warehouse field is in deep inside the grid form.
![screen shot 2017-05-23 at 1 54 31 pm](https://cloud.githubusercontent.com/assets/5196925/26346208/1c99757a-3fc3-11e7-8cd1-b1c482591d95.png)

rmehta committed:
[wip] batch, serial number selector

pratu16x7 committed:
WIP

pratu16x7 committed:
set available qty

pratu16x7 committed:
get oldest batch in modal, insert batch items

pratu16x7 committed:
batch item distribution by oldest, serial no check

pratu16x7 committed:
[fix] test

pratu16x7 committed:
fire modal on item_code, events, validation

pratu16x7 committed:
[fix] warehouse required bug

pratu16x7 committed:
wip

pratu16x7 committed:
add warehouse selection and mapping

pratu16x7 commented:
Added warehouse selection:
![batch3](https://cloud.githubusercontent.com/assets/5196925/26392581/c2a8326e-4084-11e7-8439-2a95691cfc1f.gif)

mbauskar reviewed (changes requested):
@pratu16x7,

Serial Nos can be added multiple times, also sometime after selection of serial number the serial number is added twice.

Also, on material receipt, serial number selection should be disabled

![stock-entry](https://cloud.githubusercontent.com/assets/11224291/26668090/c72c39ba-46c5-11e7-9507-9fef54b6df2b.gif)

Available Qty in batch does not reflect in child table

![batch](https://cloud.githubusercontent.com/assets/11224291/26668092/c8f31304-46c5-11e7-8bf4-619f7709794e.gif)

pratu16x7 committed:
add serial_no_check

pratu16x7 committed:
[fix] availably qty reupdate

pratu16x7 commented:
- Added duplicate check for serial no (and removed serial no from material receipt)
![serial_no_check](https://user-images.githubusercontent.com/5196925/26863108-1e4d822e-4b6e-11e7-87a0-4e1a82c68e10.gif)

- fixed available qty re-update
![serial_no_check](https://user-images.githubusercontent.com/5196925/26863179-93e15d76-4b6e-11e7-98db-e3d10ca80428.gif)

pratu16x7 committed:
cleanup

pratu16x7 committed:
[minor]

rmehta committed:
Merge branch 'develop' into serial-batch-entry

rmehta commented on erpnext/stock/doctype/stock_entry/stock_entry.py:
```diff
@@ -83,7 +83,9 @@ def set_transfer_qty(self):
 			if not flt(item.conversion_factor):
 				frappe.throw(_("Row {0}: UOM Conversion Factor is mandatory").format(item.idx))
-			item.transfer_qty = flt(item.qty * item.conversion_factor, self.precision("transfer_qty", item))
+			print type(float(item.qty))
```
print.....
